src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -82,12 +83,26 @@ public ILoggerFactory LoggerFactory
             }
         }
 
-        public AzureAppConfigurationProvider(ConfigurationClient client, AzureAppConfigurationOptions options, bool optional)
+        public AzureAppConfigurationProvider(ConfigurationClient client, AzureAppConfigurationOptions options, string environmentName, bool environmentNameEnabled, bool optional)
         {
             _client = client ?? throw new ArgumentNullException(nameof(client));
             _options = options ?? throw new ArgumentNullException(nameof(options));
             _optional = optional;
 
+            if (environmentNameEnabled)
+            {
+                try
+                {
+                    string envType = Environment.GetEnvironmentVariable(RequestTracingConstants.AspNetCoreEnvironmentVariable) ??
+                                        Environment.GetEnvironmentVariable(RequestTracingConstants.DotNetCoreEnvironmentVariable);

```

Comment:
Starting with .NET 7, the precedence order of these environment variables has changed: [ASPNET-prefixed environment variable precedence](https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/7.0/environment-variable-precedence)  

The upside of following the same logic is that our environment detection behavior will be consistent with dotnet. The downside is that it would add complexity to the implementation, since we also need to detect the type of host to make this decision: [Environments](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/environments?view=aspnetcore-7.0#environments)

@jimmyca15 @zhenlan  thoughts?

Reply:
I was hoping we don't have to read from any environment variables directly but get the environment from something like `IHostEnvironment.EnvironmentName`, so we don't care which environment variable wins in which version of .NET.

Another scenario... if either of those *__ENVIRONMENT was set, they would not match "Production" for `_environmentName`, but shouldn't we consider this scenario as production?

Reply:
I'm not sure if I understand your second point, I just know the default environment is production if neither are set, but is there another meaning if either one is set? I'm also looking to see if there's a way to get an instance of IHostEnvironment at that point in the code, but I don't see how to yet.

Reply:
> the default environment is production if neither are set

Exactly. If neither is set, does our code match it as "production"?

Reply:
Yes, the aim of this section was to check if the user specified "production" as the environment name when neither variable is set https://github.com/Azure/AppConfiguration-DotnetProvider/blob/ajusupovic/environmentname-api/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationSource.cs#L46

Reply:
The line you referenced is _not_ "neither is set". It's "either is not accessible".

Reply:
Ah I misunderstood thanks, I'll add that case

Reply:
This is one of the reasons we should avoid reading from environment variables.

Reply:
I don't think there is a way to access that `IHostEnvironment.EnvironmentName` value or any other value for the environment name given only the `IConfigurationBuilder`, which is all we have access to. All I can find is that the user could pass in the `WebHostBuilderContext` parameter for `ConfigureAppConfiguration` to `AddAzureAppConfiguration` as well and we could get the environment from there, similar to how it was described in the Background section in #312. This would mean passing in the `environmentName` parameter as well as the hosting context to `AddAzureAppConfiguration` for this new API. Does this sound like a better solution?

Reply:
We dont have access to host configuration. Users will have to pass it to us if we want to avoid reading environment variables (which might not be a bad idea). In fact, it might be better to explicitly make users pass their current environment name instead of reading environment variables. 
We could modify the new API implemented in the PR to accept a second argument that corresponds to the current environment.
```cs
public static IConfigurationBuilder AddAzureAppConfiguration( 
            this IConfigurationBuilder configurationBuilder, 
            string enabledForEnvironment, 
            string currentEnvironment, 
            Action<AzureAppConfigurationOptions> action, 
            bool optional = false) 
```

Usage would look like this:

```cs
.ConfigureAppConfiguration((hostingContext, config) => 
{
    config.AddAzureAppConfiguration(
    enabledForEnvironment: "Production",
    currentEnvironment: hostingContext.HostingEnvironment.EnvironmentName,
    options => 
    { 
        options.Connect(prod_endpoint, prod_creds) 
               .Select(KeyFilter.Any, LabelFilter.Null); 
    }); 
} 
```

Reply:
This comes back to the value of this API.
- This API is essentially a `if... else...` clause, so it's value is not about saving customer code. When a customer knows the hosting environment and what they want, it's not a problem to do the `if... else...` by themselves.
- This API's true value is relieving a customer from learning all those concepts. The API is environment-aware and does the right thing automatically.

Maybe it's okay to pass in `hostingContext`.

Reply:
Did you mean passing in `hostingContext` instead of `hostingContext.HostingEnvironment.EnvironmentName`? I feel like that would make sense because we at least do some of the work of getting the environment name. Otherwise I'm also worried that doing an if/else for the user is not that much of an improvement.

Reply:
Yes, something like that. But before we do that, it will be good to see what the code looks like for .NET/ASP.NET 6/7.

Reply:
```
public IConfigurationProvider Build(IConfigurationBuilder builder)
{
    IEnumerable<IConfigurationSource> sourceList = builder.Sources.Where(item => item is ChainedConfigurationSource);
    string currentEnvironment = null;
    if (sourceList != null && sourceList.Any())
    {
        ChainedConfigurationSource source = (ChainedConfigurationSource)sourceList.First();
        ConfigurationRoot root = (ConfigurationRoot)source.Configuration;
        if (root != null && root.Providers != null && root.Providers.Any())
        {
            ConfigurationProvider environmentVariables = (ConfigurationProvider)root.Providers.ElementAt(0);
            Console.WriteLine(environmentVariables.TryGet("ENVIRONMENT", out currentEnvironment));
        }
    }

    if (!string.IsNullOrEmpty(_environmentName))
    {
        if (string.IsNullOrEmpty(currentEnvironment))
        {
            if (!_environmentName.Equals("Production"))
            {
                return new EmptyConfigurationProvider();
            }
        }
        else
        {
            if (!currentEnvironment.Equals(_environmentName))
            {
                return new EmptyConfigurationProvider();
            }
        }
    }
```

Seems like this works to directly access the environment name from the `IConfigurationBuilder`, although it does look a bit ugly and I'm not sure if all of the ways to build an ASP.NET app would have a `ChainedConfigurationSource` yet. It did work for the ConfigStoreDemo in the examples folder from what I can tell.

Reply:
Also it seems like for .NET and ASP.NET 6/7 the main thing I see is that environment variables are managed using [ConfigurationManager](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.configurationmanager?view=dotnet-plat-ext-8.0) and are used [like this](https://github.com/dotnet/aspnetcore/blob/7264c87d64fce4420b382ddf9b834dbc97dd88d2/src/DefaultBuilder/src/WebApplicationBuilder.cs#L93) in `WebApplicationBuilder`. They just call `AddEnvironmentVariables` with a prefix and it overrides the "ENVIRONMENT" variable based on whatever is last, so "DOTNET" is prioritized in this case.